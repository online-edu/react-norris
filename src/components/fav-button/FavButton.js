import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * FavButton component.
 */
class FavButton extends Component {
  /**
   * FavButton props types.
   */
  static get propTypes() {
    return {
      /** Get fav ot not fav flag */
      favorite: PropTypes.bool.isRequired,
      /** Gets called when the fav button is clicked. */
      onClick: PropTypes.func.isRequired,
    };
  }

  /**
   * Intilalizes the state and binds all methods.
   *
   * @param {Object} props - access 'props'
   */
  constructor(props) {
    super(props);
    const { favorite } = this.props;
    this.toggle = favorite;
    this.state = {
      icon: favorite ? 'favorite' : 'favorite_border',
    };
    this.onToggle = this.onToggle.bind(this);
  }

  /**
   * Toggle icon for button and notify parent.
   */
  onToggle() {
    this.setState(
      {
        icon: this.toggle ? 'favorite_border' : 'favorite',
      },
      () => {
        this.toggle = !this.toggle;
        const { onClick } = this.props;
        onClick(this.toggle);
      },
    );
  }

  /**
   * Render method for component
   */
  render() {
    const { icon } = this.state;
    return (
      <button type="button" className="btn btn-link" onClick={this.onToggle}>
        <i className="material-icons pt-2">{icon}</i>
      </button>
    );
  }
}

export default FavButton;
