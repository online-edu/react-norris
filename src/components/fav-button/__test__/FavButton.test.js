import React from 'react';
import { shallow, mount } from 'enzyme';
import FavButton from '../FavButton';

describe('<FavButton />', () => {
  const onFavClick = jest.fn();
  const wrapper = <FavButton favorite onClick={onFavClick} />;

  it('renders correctly', () => {
    const component = shallow(wrapper);
    expect(component).toMatchSnapshot();
  });

  it('contains class name "btn"', () => {
    const component = shallow(wrapper);
    expect(component.find('.btn')).toHaveLength(1);
  });

  it('allows us to set props', () => {
    const component = mount(wrapper);
    expect(component.props().favorite).toBe(true);
    component.setProps({ favorite: false });
    expect(component.props().favorite).toBe(false);
  });

  it('simulates click events', () => {
    const component = shallow(wrapper);
    component.find('button').simulate('click');
    expect(onFavClick).toHaveBeenCalledTimes(1);
  });
});
