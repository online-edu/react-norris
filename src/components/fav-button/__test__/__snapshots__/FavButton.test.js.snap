// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavButton /> renders correctly 1`] = `ShallowWrapper {}`;
